### PR TITLE
Support Sparse CSR transpose. Fix clang-tidy warnings.

### LINKED ISSUE #61059: Transposition of CSR matrices
## 🚀 Feature
This issue describes and tracks the development of faster transposition kernels for SparseCSR on both CPU and CUDA.

## Motivation

Transposition is a fundamental operation that is often required to speed up matrix multiplication, the backwards pass of matrix multiplication or for bidirectional message passing flow in GNNs. With CSR matrices being applicable to GNNs (e.g. pytoch_sparse) and Transformers (e.g. sparse attention), which are both popular models, we should provide a fast implementation for this operation.

Transpose is a view operation in PyTorch. We currently don't have metadata extensions for CSR to represent transpositions as inexpensive metadata operations such as a CSC tensor or a "transpose" flag. We must implement transposition as a view as it will otherwise yield issues in subsequent autograd operations. (In fact, COO is currently implemented as a copy ([clone() + transpose_](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1697-L1700)), which should be a bug).

EDIT: The main usecase for fast transposition, for now, is the backwards pass of sparse matmul. We have plenty of work to do for sparse matmul and should finish it up first before we continue on with this work. This will then also give us a concrete use site and opportunity to benchmark improvements within transposition.

## Pitch

Implement rudimentary version of [```transpose.int```](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1689-L1713), [```transpose_```](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1665-L1687), [```t```](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1728-L1731) and [```t_```](https://github.com/pytorch/pytorch/blob/c50c77b444812d6f28bb7a9071ca0eb643ab6f78/aten/src/ATen/native/TensorShape.cpp#L1733-L1736) for SparseCsrCPU and SparseCsrCUDA Tensors by modifying the crow_indices and col_indices only and creating a view of the values. Write tests to confirm correctness of forward and autograd. Write benchmarks to establish baseline performance and investigate options for speedup.

EDIT: CSR metadata extensions are mandatory and simply modifying crow and col indicies won't be enough. See discussion below.

## Alternatives

Currently this gap is closed by one of our partner libraries [pytorch_sparse](https://github.com/rusty1s/pytorch_sparse).

## Additional context

```
import torch
crow_indices = torch.tensor([0, 2, 4])
col_indices = torch.tensor([0, 1, 0, 1])
values = torch.tensor([1, 2, 3, 4], dtype=torch.float)
tensor0 = torch.sparse_csr_tensor(crow_indices, col_indices, values)
tensor0.transpose(0, 1)
```

yields

```
Traceback (most recent call last):
  File "/data/home/cpuhrsch/tmp/csr_transpose_bench.py", line 7, in <module>
    tensor0.transpose(0, 1)
NotImplementedError: Could not run 'aten::as_strided' with arguments from the 'SparseCsrCPU' backend
```

Similarly for CUDA we have

```
import torch

crow_indices = torch.tensor([0, 2, 4])
col_indices = torch.tensor([0, 1, 0, 1])
values = torch.tensor([1, 2, 3, 4], dtype=torch.float)
tensor0 = torch.sparse_csr_tensor(crow_indices, col_indices, values, device=torch.device('cuda'))
tensor0.transpose(0, 1)
```

yields

```
Traceback (most recent call last):
  File "/data/home/cpuhrsch/tmp/csr_transpose_bench.py", line 7, in <module>
    tensor0.transpose(0, 1)
NotImplementedError: Could not run 'aten::as_strided' with arguments from the 'SparseCsrCUDA' backend.
```

cc @albanD @IvanYashchuk @pearu @mruberry @bhosmer

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #70582
* #70581
* #70580

cc @nikitaved @pearu @cpuhrsch

Differential Revision: [D33414446](https://our.internmc.facebook.com/intern/diff/D33414446)

### CODE REVIEW

**Comment by pearu** [comment]
This PR replaces https://github.com/pytorch/pytorch/pull/68883 and https://github.com/pytorch/pytorch/pull/68938

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -1650,6 +1685,14 @@ Tensor transpose(const Tensor & self, int64_t dim0, int64_t dim1) {
   auto ndims = self.dim();
   dim0 = maybe_wrap_dim(dim0, ndims);
   dim1 = maybe_wrap_dim(dim1, ndims);
+
+  if (self.is_sparse_csr()) {
+    // Sparse CSR transpose is a copy operation as the values of
+    // transposed CSR tensor are permuted values of the input CSR
+    // tensor.
```
Comment (cpuhrsch) [inline aten/src/ATen/native/TensorShape.cpp:1692]: NIT: You could add this to the other sparse branch as well or unify them and then dispatch to COO or CSR as needed.

**Review by cpuhrsch — APPROVED** [review]

**Comment by cpuhrsch** [comment]
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33414446).

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -1562,6 +1562,41 @@ static inline Tensor & sparse_transpose_(Tensor & self, int64_t dim0, int64_t di
   return self;
 }
 
+static inline Tensor sparse_csr_transpose(const Tensor & self, int64_t dim0, int64_t dim1) {
+  TORCH_INTERNAL_ASSERT(self.is_sparse_csr());
+  auto dim0_ = maybe_wrap_dim(dim0, 2);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/TensorShape.cpp:1567]: These are unused which trips up a test

File: aten/src/ATen/native/TensorShape.cpp [hunk 3]
```diff
@@ -1562,6 +1562,41 @@ static inline Tensor & sparse_transpose_(Tensor & self, int64_t dim0, int64_t di
...
+  auto dim1_ = maybe_wrap_dim(dim1, 2);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/TensorShape.cpp:1568]: These are unused which trips up a test

**Review by cpuhrsch — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/TensorShape.cpp (hunk 2 above)
Comment (pearu) [reply aten/src/ATen/native/TensorShape.cpp:1567]: Fixed. In fact, sparse_csr_transpose does not need the dimensions arguments.

File: aten/src/ATen/native/TensorShape.cpp (hunk 1 above)
Comment (pearu) [reply aten/src/ATen/native/TensorShape.cpp]: Done.

File: aten/src/ATen/native/TensorShape.cpp (hunk 3 above)
Comment (pearu) [reply aten/src/ATen/native/TensorShape.cpp]: Removed dim0 and dim1 arguments.

**Comment by cpuhrsch** [comment]
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33414446).

**Review by cpuhrsch — APPROVED** [review]

**Review by cpuhrsch — APPROVED** [review]

**Pull request closed without merging**